src/engine/UNSATCertificate.h

```diff
@@ -0,0 +1,282 @@
+/*********************                                                        */
+/*! \file UNSATCertificate.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Omri Isac, Guy Katz
+ ** This file is part of the Marabou project.
+ ** Copyright (c) 2017-2022 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+ **/
+
+
+#ifndef __UNSATCertificate_h__
+#define __UNSATCertificate_h__
+
+#include "BoundExplainer.h"
+#include "SmtLibWriter.h"
+#include "PiecewiseLinearFunctionType.h"
+#include "ReluConstraint.h"
+
+/*
+  Contains all necessary info of a ground bound update during a run (i.e from ReLU phase-fixing)
+*/
+struct PLCExplanation
+{
+    unsigned _causingVar;
+    unsigned _affectedVar;
+    double _bound;
+    bool _isCausingBoundUpper;
+    bool _isAffectedBoundUpper;
+    double *_explanation;
+    PiecewiseLinearFunctionType _constraintType;
+    unsigned _decisionLevel;
+
+    ~PLCExplanation()
+    {
+        if ( _explanation )
+        {
+            delete [] _explanation;
+            _explanation = NULL;
+        }
+    }
+};
+
+/*
+  Contains all ino relevant for a simple Marabou contradiction - i.e. explanations of contradicting bounds of a variable
+*/
+struct Contradiction
+{
+    unsigned _var;
+    double *_upperBoundExplanation;
+    double *_lowerBoundExplanation;
+
+    ~Contradiction()
+    {
+        if ( _upperBoundExplanation )
+        {
+            delete [] _upperBoundExplanation;
+            _upperBoundExplanation = NULL;
+        }
+
+        if ( _lowerBoundExplanation )
+        {
+            delete [] _lowerBoundExplanation;
+            _lowerBoundExplanation = NULL;
+        }
+    }
+};
+
+/*
+  A smaller representation of a problem constraint
+*/
+struct ProblemConstraint
+{
+    PiecewiseLinearFunctionType _type;
+    List<unsigned> _constraintVars;
+    PhaseStatus _status;
+
+    bool operator==( const ProblemConstraint other ) const
+    {
+        return _type == other._type && _constraintVars == other._constraintVars;
+    }
+};
+
+enum DelegationStatus : unsigned
+{
+    DONT_DELEGATE = 0,
+    DELEGATE_DONT_SAVE = 1,
+    DELEGATE_SAVE = 2
+};
+
+/*
+  A certificate node in the tree representing the UNSAT certificate
+*/
+class CertificateNode
+{
+public:
+    /*
+      Constructor for the root
+    */
+    CertificateNode( Vector<Vector<double>> *_initialTableau, Vector<double> &groundUpperBounds, Vector<double> &groundLowerBounds );
+
+    /*
+      Constructor for a regular node
+    */
+    CertificateNode( CertificateNode *parent, PiecewiseLinearCaseSplit split );
+
+    ~CertificateNode();
+
+    /*
+      Certifies the tree is indeed a correct proof of unsatisfiability;
+    */
+    bool certify();
```

Comment:
Feels a bit strange that this is here. Perhaps we should have a checker class?

Reply:
The certification process heavily relies on the internals of CertificateNode. It is possible to have another class, but then I'm not sure code will be cleaner (even if it'll be a friend class). I also find it elegant that we encapsulate everything in a single method, and don't need to add external objects when certifying.